File: uis/client/mobile_app/lib/services/loan_service.dart
import 'api_service.dart';
import 'error_handler_service.dart';
import '../config/app_config.dart';

class LoanService {
  final ApiService _apiService;

  LoanService(this._apiService);

  // =================== APPLY FOR LOAN ===================
  Future<Map<String, dynamic>> applyForLoan({
    required double amount,
    required int termInMonths,
    required String purpose,
    double monthlyIncome = 1000,
    double existingDebt = 0,
    double collateralValue = 5000,
    int creditScore = 1100,
    String employmentStatus = "employed",
    int employmentDuration = 24,
    int bankStatementScore = 90,
    bool bvnVerified = true,
    bool ninVerified = true,
  }) async {
    try {
      final response = await _apiService.post(
        '${AppConfig.loanEndpoint}/loans/applications',
        data: {
          "loan_amount": amount,
          "loan_purpose": purpose,
          "requested_term": termInMonths,
          "monthly_income": monthlyIncome,
          "existing_debt": existingDebt,
          "collateral_value": collateralValue,
          "credit_score": creditScore,
          "employment_status": employmentStatus,
          "employment_duration": employmentDuration,
          "bank_statement_score": bankStatementScore,
          "bvn_verified": bvnVerified,
          "nin_verified": ninVerified,
        },
      );

      if (response.data['success'] == true || response.statusCode == 200) {
        return {
          'success': true,
          'message': response.data['message'] ?? 'Loan application submitted',
          'data': response.data['data'],
        };
      } else {
        return {
          'success': false,
          'message': ErrorHandlerService.handleError(response.data),
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(e),
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'An unexpected error occurred while applying for loan.',
      };
    }
  }

  // =================== GET ALL LOANS ===================
  /// Matches web app: ${AppConfig.loanEndpoint}/applications
  Future<List<Map<String, dynamic>>> getAllLoans() async {
    try {
      // Web app endpoint: ${AppConfig.loanEndpoint}/applications
      final response = await _apiService.get(
        '${AppConfig.loanEndpoint}/loans/applications',
      );

      // Web app: response.data is the array directly, not wrapped in {success, data}
      if (response.statusCode == 200) {
        // Response is an array directly: [{...}, {...}]
        if (response.data is List) {
          return (response.data as List)
              .map((loan) => loan as Map<String, dynamic>)
              .toList();
        }
        // If wrapped in data field
        if (response.data is Map && response.data['data'] is List) {
          return (response.data['data'] as List)
              .map((loan) => loan as Map<String, dynamic>)
              .toList();
        }
        return [];
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  // =================== FETCH LOAN DETAILS ===================
  Future<Map<String, dynamic>> fetchLoanDetails(String loanId) async {
  final response = await _apiService.get(
    '${AppConfig.loanEndpoint}/loans/applications/$loanId',
  );

  final data = response.data;
  print("FETCH LOAN DETAILS RESPONSE: $data");

  if (response.statusCode == 200 ) {
    return Map<String, dynamic>.from(data);
  }

  throw Exception('Loan not found');
}
  // Get loan repayment schedule
  Future<Map<String, dynamic>> getRepaymentSchedule(String loanId) async {
    try {
      final response = await _apiService.get(
        '${AppConfig.loanEndpoint}/loans/$loanId/schedule',
      );

      if (response.data['success'] == true || response.statusCode == 200) {
        return {
          'success': true,
          'data': response.data['data'],
        };
      } else {
        return {
          'success': false,
          'message': ErrorHandlerService.handleError(response.data),
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(e),
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'An unexpected error occurred while fetching repayment schedule.',
      };
    }
  }
}

File: uis/client/web2/src/services/loan_service.ts
import { AppConfig } from '../config/app_config';
import { getErrorMessage } from '../utils/error_utils';
import { apiService } from './api_service';

export class LoanService {
  // =================== APPLY FOR LOAN ===================
  async applyForLoan(params: {
    amount: number;
    termInMonths: number;
    purpose: string;
    monthlyIncome?: number;
    existingDebt?: number;
    collateralValue?: number;
    creditScore?: number;
    employmentStatus?: string;
    employmentDuration?: number;
    bankStatementScore?: number;
    bvnVerified?: boolean;
    ninVerified?: boolean;
  }): Promise<Record<string, any>> {
    try {
      const response = await apiService.post(`${AppConfig.loanEndpoint}/applications`, {
        loan_amount: params.amount,
        loan_purpose: params.purpose,
        requested_term: params.termInMonths,
        monthly_income: params.monthlyIncome ?? 1000,
        existing_debt: params.existingDebt ?? 0,
        collateral_value: params.collateralValue ?? 5000,
        credit_score: params.creditScore ?? 1100,
        employment_status: params.employmentStatus ?? 'employed',
        employment_duration: params.employmentDuration ?? 24,
        bank_statement_score: params.bankStatementScore ?? 90,
        bvn_verified: params.bvnVerified ?? true,
        nin_verified: params.ninVerified ?? true,
      });

      if (response.status === 200 || response.status === 201) {
        return response.data as Record<string, any>;
      } else {
        const data = response.data as { message?: string };
        throw new Error(getErrorMessage(data, 'Loan application failed'));
      }
    } catch (error: unknown) {
      throw new Error(getErrorMessage(error, 'Loan application failed'));
    }
  }

  // =================== FETCH LOAN DETAILS ===================
  async fetchLoanDetails(loanId: string): Promise<Record<string, any>> {
    try {
      const response = await apiService.get(`${AppConfig.loanEndpoint}/application/${loanId}`);
      if (response.status === 200) {
        const data = response.data as { data?: any; message?: string };
        return data.data;
      } else {
        const data = response.data as { message?: string };
        throw new Error(getErrorMessage(data, 'Failed to fetch loan details'));
      }
    } catch (error: unknown) {
      throw new Error(getErrorMessage(error, 'Failed to fetch loan details'));
    }
  }

  // =================== GET ALL LOANS ===================
  async getAllLoans(): Promise<any[]> {
    try {
      const response = await apiService.get(`${AppConfig.loanEndpoint}/applications`);
      if (response.status === 200) {
        return (response.data as any[]) || [];
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch loans:', error);
      return [];
    }
  }

  // =================== GET LOAN BY ID ===================
  async getLoanById(loanId: string): Promise<any> {
    try {
      const response = await apiService.get(`${AppConfig.loanEndpoint}/applications/${loanId}`);
      if (response.status === 200) {
        return response.data as Record<string, any>;
      }
      const data = response.data as { message?: string };
      throw new Error(getErrorMessage(data, 'Loan not found'));
    } catch (error: unknown) {
      throw new Error(getErrorMessage(error, 'Failed to fetch loan details'));
    }
  }

  // =================== REPAY LOAN ===================
  async repayLoan(loanId: string, amount: number): Promise<{ success: boolean; message: string }> {
    try {
      const response = await apiService.post(`${AppConfig.loanEndpoint}/${loanId}/repay`, {
        amount,
      });

      const data = response.data as { message?: string };
      if (response.status === 200) {
        return {
          success: true,
          message: getErrorMessage(data, 'Loan repayment successful'),
        };
      } else {
        return {
          success: false,
          message: getErrorMessage(data, 'Loan repayment failed'),
        };
      }
    } catch (error: unknown) {
      return {
        success: false,
        message: getErrorMessage(error, 'Loan repayment error'),
      };
    }
  }

  // =================== GET REPAYMENT SCHEDULE ===================
  async getRepaymentSchedule(loanId: string): Promise<any[]> {
    try {
      const response = await apiService.get(`${AppConfig.loanEndpoint}/loans/${loanId}/schedule`);
      if (response.status === 200) {
        return (response.data as any[]) || [];
      }
      return [];
    } catch (error) {
      console.error('Failed to fetch repayment schedule:', error);
      return [];
    }
  }
}

export const loanService = new LoanService();
